components/net/bluetooth_thread.rs

```diff
@@ -472,35 +477,43 @@ impl BluetoothManager {
     fn gatt_server_connect(&mut self, device_id: String, sender: IpcSender<BluetoothResult<bool>>) {
         let mut adapter = get_adapter_or_return_error!(self, sender);
 
-        let connected = match self.get_device(&mut adapter, &device_id) {
+        match self.get_device(&mut adapter, &device_id) {
             Some(d) => {
                 if d.is_connected().unwrap_or(false) {
-                    true
-                } else {
-                    d.connect().is_ok()
+                    return drop(sender.send(Ok(true)));
                 }
+                let _ = d.connect();
+                for _ in 0..MAXIMUM_TARNSACTION_TIME {
+                    match d.is_connected().unwrap_or(false) {
+                        true => return drop(sender.send(Ok(true))),
+                        false => thread::sleep(Duration::from_millis(CONNECTION_TIMEOUT_MS)),
+                    }
+                }
+                return drop(sender.send(Err(String::from(NETWORK_ERROR))));
             },
             None => return drop(sender.send(Err(String::from(DEVICE_ERROR)))),
-        };
-
-        let _ = sender.send(Ok(connected));
+        }
     }
 
     fn gatt_server_disconnect(&mut self, device_id: String, sender: IpcSender<BluetoothResult<bool>>) {
         let mut adapter = get_adapter_or_return_error!(self, sender);
 
-        let connected = match self.get_device(&mut adapter, &device_id) {
+        match self.get_device(&mut adapter, &device_id) {
             Some(d) => {
-                if d.is_connected().unwrap_or(false) {
-                    d.disconnect().is_ok()
-                } else {
-                    false
+                if !d.is_connected().unwrap_or(true) {
+                    return drop(sender.send(Ok(false)));
                 }
+                let _ = d.disconnect();
+                for _ in 0..MAXIMUM_TARNSACTION_TIME {
+                    match d.is_connected().unwrap_or(true) {
+                        false => return drop(sender.send(Ok(false))),
```

Comment:
if you want to keep the drop/sleep order, negate the expr like in [L503](https://github.com/szeged/servo/pull/42/files#diff-438a3422f5ba8c39dabd8e4e18928c51R503)